File: DP-3-wayToSchool/wayToSchool.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles)
{
    long long d[n + 1][m + 1];
    memset(d, 0, sizeof(d));
    d[0][1] = 1;

    for (int k = 0; k < puddles.size(); k++)
    {
        d[puddles[k][1]][puddles[k][0]] = -1;
    }

    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= m; j++)
        {
            if (d[i][j] == -1)
                d[i][j] = 0;
            else
                d[i][j] = (d[i - 1][j] + d[i][j - 1]) % 1000000007;
        }
    }
    return d[n][m];
}

int main()
{
    // return 4
    // cout << solution(4, 3, {{2, 2}}) << endl;
    cout << solution(100, 100, {{2, 2}}) << endl;
    // cout << solution(4, 3, {{2, 2}}) << endl;

    return 0;
}

File: DP-3-wayToSchool/wayToSchool.py
def solution(m, n, puddles):
    d = [[0] * m for _ in range(n)]
    for c, r in puddles:
        d[r-1][c-1] = -1
    for c in range(1, m):
        if d[0][c] != -1:
            d[0][c] = 1
        else:
            break
    for r in range(1, n):
        if d[r][0] != -1:
            d[r][0] = 1
        else:
            break
    for r in range(1, n):
        for c in range(1, m):
            if d[r][c] == -1:
                continue
            if d[r-1][c] == -1 and d[r][c-1] == -1:
                d[r][c] = 0
            elif d[r-1][c] == -1 and d[r][c-1] != -1:
                d[r][c] = d[r][c-1]
            elif d[r-1][c] != -1 and d[r][c-1] == -1:
                d[r][c] = d[r-1][c]
            else:
                d[r][c] = d[r-1][c] + d[r][c-1]
    return d[-1][-1] % 1000000007


print(solution(4, 3, [[2, 2]]))
# return 4
